# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/credentialmanagement/identity/NavigatorLogin.cpp
# Commit: 459739221a88
# Full Hash: 459739221a884b2b1c2adab42fd19fafe279bd20
# Author: Sandor Molnar <smolnar@mozilla.com>
# Date: 2025-06-05 04:07:37
# Regressor Bug: 1969017
# File Overlap Count: 2
# Description:
#   Revert "Bug 1969015, 1969017: apply code formatting via Lando" for causing build bustages
#   
#   This reverts commit ab681e36aeafc21ba62bb4680bdaa6206f02546d.
#   
#   Revert "Bug 1969017 - Refactor FedCM to use a managed actor rather than tagging onto WindowGlobalParent - r=anti-tracking-reviewers,emz"
# ==============================================================================

diff -r fc9986230203 -r 459739221a88 dom/credentialmanagement/identity/NavigatorLogin.cpp
--- a/dom/credentialmanagement/identity/NavigatorLogin.cpp	Thu Jun 05 00:15:19 2025 +0300
+++ b/dom/credentialmanagement/identity/NavigatorLogin.cpp	Thu Jun 05 00:19:41 2025 +0300
@@ -11,7 +11,7 @@
 #include "mozilla/dom/Promise.h"
 #include "mozilla/Maybe.h"
 #include "mozilla/net/SFVService.h"
-#include "mozilla/dom/WebIdentityHandler.h"
+#include "mozilla/dom/WindowGlobalChild.h"
 #include "nsCycleCollectionParticipant.h"
 #include "nsIGlobalObject.h"
 #include "nsIPermissionManager.h"
@@ -57,30 +57,54 @@
   return NavigatorLogin_Binding::Wrap(aCx, this, aGivenProto);
 }
 
-NavigatorLogin::NavigatorLogin(nsPIDOMWindowInner* aGlobal) : mOwner(aGlobal) {
+NavigatorLogin::NavigatorLogin(nsIGlobalObject* aGlobal) : mOwner(aGlobal) {
   MOZ_ASSERT(mOwner);
 };
 
 already_AddRefed<mozilla::dom::Promise> NavigatorLogin::SetStatus(
     LoginStatus aStatus, mozilla::ErrorResult& aRv) {
-  RefPtr<Promise> promise = Promise::Create(mOwner->AsGlobal(), aRv);
+  RefPtr<Promise> promise = Promise::Create(mOwner, aRv);
   if (aRv.Failed()) {
     return nullptr;
   }
 
-  if (!CredentialsContainer::IsSameOriginWithAncestors(mOwner)) {
+  nsPIDOMWindowInner* window = mOwner->GetAsInnerWindow();
+  if (!window) {
+    promise->MaybeRejectWithUnknownError(
+        "navigator.login.setStatus called on unavailable window"_ns);
+    return promise.forget();
+  }
+
+  if (!CredentialsContainer::IsSameOriginWithAncestors(window)) {
     promise->MaybeRejectWithSecurityError(
         "navigator.login.setStatus must be called in a frame that is same-origin with its ancestors"_ns);
     return promise.forget();
   }
 
-  WebIdentityHandler* identityHandler = mOwner->GetOrCreateWebIdentityHandler();
-  if (!identityHandler) {
-    promise->MaybeRejectWithOperationError("");
+  WindowGlobalChild* wgc = window->GetWindowGlobalChild();
+  if (!wgc) {
+    promise->MaybeRejectWithUnknownError(
+        "navigator.login.setStatus called while window already destroyed"_ns);
     return promise.forget();
   }
 
-  identityHandler->SetLoginStatus(aStatus, promise);
+  wgc->SendSetLoginStatus(aStatus)->Then(
+      GetCurrentSerialEventTarget(), __func__,
+      [promise](
+          const WindowGlobalChild::SetLoginStatusPromise::ResolveValueType&
+              aResult) {
+        if (NS_SUCCEEDED(aResult)) {
+          promise->MaybeResolveWithUndefined();
+        } else {
+          promise->MaybeRejectWithUnknownError(
+              "navigator.login.setStatus had an unexpected internal error");
+        }
+      },
+      [promise](const WindowGlobalChild::SetLoginStatusPromise::RejectValueType&
+                    aResult) {
+        promise->MaybeRejectWithUnknownError(
+            "navigator.login.setStatus had an unexpected internal error");
+      });
   return promise.forget();
 }
 